### PR TITLE
delay compilation of training functions until they are needed

### PR DESCRIPTION
**Pull request by mbrookhart**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/ngraph/ngraph_graph.h [hunk 1]
```diff
@@ -219,6 +219,7 @@ class Graph : public Node {
       ngraph_forward[i] = nullptr;
       ngraph_backward[i] = nullptr;
     }
+    fprop_cache = ngraph::FpropCache();
```
Comment (cconvey) [inline src/ngraph/ngraph_graph.h]: Is this the last time we'll actually want a populated `fprop_cache` for this `ngraph_bridge::Graph` object?

If so, I wonder if it would be a little more intuitive to have `fprop_cache` be a `shared_ptr<ngraph::FpropCache>`, and we'd replace this statement with setting the pointer to null?

File: src/ngraph/ngraph_nnvm_ops.cc [hunk 2]
```diff
@@ -77,6 +78,16 @@ void update_aux_vals(const std::shared_ptr<Graph> &graph,
   }
 }
 
+void compile_if_needed(std::shared_ptr<Graph> graph, int mode) {
+  if (mode == static_cast<int>(GraphExeMode::kTrain)) {
```
Comment (cconvey) [inline src/ngraph/ngraph_nnvm_ops.cc]: Any reason to not use the `int(GraphExeMode::kTrain)`?  It strikes me as more succinct and a tad easier to read.

Comment (mbrookhart) [reply src/ngraph/ngraph_nnvm_ops.cc]: Do you mean (int)(GraphExeMode::kTrain)? If so, it's less safe, in general, we shouldn't be using C-style casting.

Comment (cconvey) [reply src/ngraph/ngraph_nnvm_ops.cc]: Nope, I really did mean `int(GraphExeMode::kTrain)`.

I'm not aware of any safety concerns when the operand is a enum-class literal.  But I wouldn't recommend holding up the PR just because of this either.

Comment (mbrookhart) [reply src/ngraph/ngraph_nnvm_ops.cc]: ...I'm not sure I've ever seen int as a function in C++.

File: src/ngraph/ngraph_graph.h [hunk 3]
```diff
@@ -219,6 +219,7 @@ class Graph : public Node {
       ngraph_forward[i] = nullptr;
       ngraph_backward[i] = nullptr;
     }
+    fprop_cache = nullptr;
```
Comment (cconvey) [inline src/ngraph/ngraph_graph.h]: Just FYI in case you didn't know: the default c'tor for `shared_ptr` will initialize to nullptr.

So the `= nullptr` above isn't technically necessary.

File: src/ngraph/ngraph_nnvm_ops.cc [hunk 4]
```diff
@@ -77,6 +78,16 @@ void update_aux_vals(const std::shared_ptr<Graph> &graph,
...
+    if (graph->ngraph_forward[mode] == nullptr) {
+      CompileForwardBackward(graph, graph->fprop_cache->fprop,
```
Comment (cconvey) [inline src/ngraph/ngraph_nnvm_ops.cc]: Is it possible for `graph->fprop_cache` to ever be null when this code runs?

File: src/ngraph/ngraph_sgcompiler.cc [hunk 5]
```diff
@@ -342,9 +340,10 @@ void SGCompiler::CompileSubgraph(std::shared_ptr<Graph> sub_graph) {
   }
 
   if (exe_mode_ == GraphExeMode::kTrain) {
-    ngraph::FpropCache fprop_cache;
-    fprop_cache.node_param_map = std::make_shared<ngraph::NodeMap>();
-    CompileForwardBackward(sub_graph, f, maybe_bf, exe_mode_, fprop_cache);
+    sub_graph->fprop_cache->fprop = f;
+    sub_graph->fprop_cache->bprop = maybe_bf;
+    sub_graph->fprop_cache->node_param_map =
+        std::make_shared<ngraph::NodeMap>();
```
Comment (cconvey) [inline src/ngraph/ngraph_sgcompiler.cc]: Is it possible for `sub_graph->fprop_cache` to ever be null when this code runs?

File: src/ngraph/ngraph_graph.h (hunk 3 above)
Comment (mbrookhart) [reply src/ngraph/ngraph_graph.h]: This is deleting it :)

File: src/ngraph/ngraph_sgcompiler.cc (hunk 5 above)
Comment (mbrookhart) [reply src/ngraph/ngraph_sgcompiler.cc]: Yes! It didn't used to be, but it now is. Will fix.

File: src/ngraph/ngraph_graph.h (hunk 3 above)
Comment (cconvey) [reply src/ngraph/ngraph_graph.h]: D'oh, that's what I get for reviewing a PR while listening to Jason's presentation.  Sorry :/
